lib/galaxy/tools/parameters/basic.py

```diff
@@ -1433,18 +1435,23 @@ def from_json(self, value, trans, other_values={}):
             if self.optional:
                 return None
             raise ValueError("parameter '%s': an invalid option (%r) was selected" % (self.name, value))
-        elif not legal_values:
+        elif not legal_values and not self.is_dynamic:
             raise ValueError("parameter '%s': requires a value, but no legal values defined" % (self.name))
+        if isinstance(value, dict):
+            clicked_node = value.get('last_clicked_node')
```

Comment:
I'm really nervous about this complexity, in this file in particular and without test cases. @nuwang and I discussed adding WebDav to the upload component and rule builder via a plugin framework as an alternative to this. I will try to prototype this... I hope...